clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1011,8 +996,14 @@ private Record nextFetchedRecord() {
                         drain();
                         return null;
                     }
-                    currentBatch = batches.next();
-                    maybeEnsureValid(currentBatch);
+
+                    if (hasExceptionInLastFetch) {
+                        currentBatch.ensureValid();
+                    } else {
+                        lastRecord = null;
+                        currentBatch = batches.next();
+                        maybeEnsureValid(currentBatch);
```

Comment:
Wasn't sure if you were pushing another change, but I'll go and leave a comment. I'm unclear what should happen if the checksum fails on the batch. It seems like we would still catch the exception and set `hasExceptionInLastFetch` to true. But then we would fail because we are trying to call `maybeEnsureValid` on `lastRecord` which is now null.